hwloc/topology-rsmi.c

```diff
@@ -0,0 +1,268 @@
+    rsmi_rc = rsmi_status_string(rsmi_rc, &status_string);
+    fprintf(stderr, "Failed to get Unique ID of the GPU: %s\n", status_string);
+    return;
+  }
+  sprintf(buffer, "%lx", id);
+}
+
+/*
+ * Get the serial number of the GPU
+ *
+ * dv_ind  (IN) The device index
+ * serial  (OUT) Serial number of GPU devices
+ * size    (IN) Length of the caller provided buffer
+ */
+static void get_device_serial_number(uint32_t dv_ind, char *serial, unsigned int size)
+{
+  const char *status_string;
+  rsmi_status_t rsmi_rc = rsmi_dev_serial_number_get(dv_ind, serial, size);
+
+  if (rsmi_rc != RSMI_STATUS_SUCCESS) {
+    rsmi_rc = rsmi_status_string(rsmi_rc, &status_string);
+    fprintf(stderr, "Failed to get serial_number of the GPU: %s\n", status_string);
+    serial[0] = '\0';
+  }
+}
+
+static int
+hwloc_rsmi_discover(struct hwloc_backend *backend, struct hwloc_disc_status *dstatus)
+{
+  /*
+   * This backend uses the underlying OS.
+   * However we don't enforce topology->is_thissystem so that
+   * we may still force use this backend when debugging with !thissystem.
+   */
+
+  struct hwloc_topology *topology = backend->topology;
+  enum hwloc_type_filter_e filter;
+  rsmi_status_t ret;
+  unsigned nb, i;
+
+  assert(dstatus->phase == HWLOC_DISC_PHASE_IO);
+
+  hwloc_topology_get_type_filter(topology, HWLOC_OBJ_OS_DEVICE, &filter);
+  if (filter == HWLOC_TYPE_FILTER_KEEP_NONE)
+    return 0;
+
+  rsmi_init(0);
+
+  ret = rsmi_num_monitor_devices(&nb);
+  if (RSMI_STATUS_SUCCESS != ret || !nb) {
+    if (RSMI_STATUS_SUCCESS != ret)
+      fprintf(stderr, "Failed to get device count\n");
+    rsmi_shut_down();
+    return 0;
+  }
+
+  for (i=0; i<nb; i++) {
+    rsmiPciInfo_t pci;
+    hwloc_obj_t osdev, parent;
+    char buffer[64];
+
+    osdev = hwloc_alloc_setup_object(topology, HWLOC_OBJ_OS_DEVICE, HWLOC_UNKNOWN_INDEX);
+    snprintf(buffer, sizeof(buffer), "rsmi%u", i);
+    osdev->name = strdup(buffer);
+    osdev->depth = HWLOC_TYPE_DEPTH_UNKNOWN;
+    osdev->attr->osdev.type = HWLOC_OBJ_OSDEV_GPU;
+
+    hwloc_obj_add_info(osdev, "Backend", "RSMI");
+    hwloc_obj_add_info(osdev, "GPUVendor", "Advanced Micro Devices");
+
+    buffer[0] = '\0';
+    get_device_name(i, buffer, sizeof(buffer));
+    hwloc_obj_add_info(osdev, "GPUModel", buffer);
+
+    buffer[0] = '\0';
+    get_device_serial_number(i, buffer, sizeof(buffer));
+    if (buffer[0] != '\0')
+      hwloc_obj_add_info(osdev, "AMDSerial", buffer);
+
+    buffer[0] = '\0';
+    get_device_unique_id(i, buffer);
+    if (buffer[0] != '\0')
+      hwloc_obj_add_info(osdev, "AMDUUID", buffer);
+
+    parent = NULL;
+    get_device_pci_info(i, &pci);
+    parent = hwloc_pci_find_parent_by_busid(topology, pci.domain, pci.bus, pci.device, 0);
+    get_device_pci_linkspeed(i, &(parent->attr->pcidev.linkspeed));
+    if (!parent)
+      parent = hwloc_get_root_obj(topology);
+
+    hwloc_insert_object_by_parent(topology, parent, osdev);
+  }
+
+  rsmi_shut_down();
```

Comment:
What happens if the program does its own rsmi_init() either earlier or before? do you have proper reference counting inside RSMI to make sure this rsmi_shut_down() in hwloc shut down RSMI for the entire process?

Reply:
The first time init() is called, the internal data structures are initialized, and the first time shutdown is called they are freed.

Reply:
There is no reference counting inside RSMI right now, can be added later.

Reply:
It *must* be added as soon as possible. Missing support for multiple init/shutdown is very annoying for developers. Right now, it means hwloc just cannot call shutdown at all: if an application (e.g using hwloc/rsmi.h) calls rsmi_init() before hwloc_topology_load(), hwloc would shutdown rsmi at the end of that hwloc_topology_load().

Not calling shutdown means valgrind will complain. Hence we'll end up adding yet another environment variable sayign "OK hwloc, you can shutdown rsmi, I know what I am doing" so that our regression tests don't fail.

Reply:
Let me talk to the developer.

Reply:
I put all details in a github issue against rsmi https://github.com/RadeonOpenCompute/rocm_smi_lib/issues/62

Reply:
I will forward it to the developer for rocm_smi library. Like I said at beginning, this is just phase 1 for hwloc plugin. It will be added in phase 2.